### PR TITLE
fix peer score accumulation

### PR DESCRIPTION
**Pull request by arnetheduck**
* fix accumulating peer score
* fix missing exception handling
* remove unnecessary initHashSet/initTable calls
* simplify peer stats management
* clean up tests a little
* fix some missing raises annotations

### CODE REVIEW

File: libp2p/protocols/pubsub/mcache.nim [hunk 1]
```diff
@@ -24,21 +26,20 @@ type
 
 func get*(c: MCache, mid: MessageID): Option[Message] =
   if mid in c.msgs:
-    some(c.msgs[mid])
+    try: some(c.msgs[mid])
+    except KeyError: raiseAssert "checked"
```
Comment (dryajov) [inline libp2p/protocols/pubsub/mcache.nim:30]: why do we want to die here?

**Review by dryajov — APPROVED** [review]
Nice cleanup, just a few comments. 


Also, curious why are you using `raiseAssert` so aggressively, specially where it seems like a missing key is pretty innocuous?

**Comment by sinkingsugar** [comment]
I must say I am happy finally someone is taking a proper look at my code, I was really starting to feel alone in this crusade.

I'm off this week but had a look anyway.

I think behaviour penalty looks different in go:
https://github.com/libp2p/go-libp2p-pubsub/blob/6c1addf493ebb942548dbe027fd1fef2e1f8391d/score.go#L326
I think that's what I was doing with the squared but probably was wrong as well.

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 2]
```diff
@@ -159,16 +159,16 @@ method init*(g: GossipSub) =
...
-    # restore previously stored score
-    let stats = g.peerStats[peer.peerId]
+  g.withPeerStats(peer.peerId) do (stats: var PeerStats):
+    # Make sure stats and peer information match, even when reloading peer stats
+    # from a previous connection
     peer.score = stats.score
     peer.appScore = stats.appScore
     peer.behaviourPenalty = stats.behaviourPenalty
 
+    peer.iWantBudget = IWantPeerBudget
```
[the remark this answers is not in the document]
Comment (arnetheduck) [reply libp2p/protocols/pubsub/gossipsub.nim:169]: the budget is reset on every heartbeat, so it doesn't really matter

File: libp2p/protocols/pubsub/gossipsub/behavior.nim [hunk 3]
```diff
@@ -400,7 +387,7 @@ proc rebalanceMesh*(g: GossipSub, topic: string, metrics: ptr MeshMetrics = nil)
 
   # opportunistic grafting, by spec mesh should not be empty...
   if g.mesh.peers(topic) > 1:
```
[the remark this answers is not in the document]
Comment (arnetheduck) [reply libp2p/protocols/pubsub/gossipsub/behavior.nim:389]: not sure, didn't really analyze that part

File: libp2p/protocols/pubsub/gossipsub/behavior.nim [hunk 4]
```diff
@@ -474,12 +461,14 @@ proc dropFanoutPeers*(g: GossipSub) =
...
-  for topic in toSeq(g.lastFanoutPubSub.keys):
-    let val = g.lastFanoutPubSub[topic]
+  var drops: seq[string]
+  for topic, val in g.lastFanoutPubSub:
     if now > val:
       g.fanout.del(topic)
-      g.lastFanoutPubSub.del(topic)
+      drops.add topic
       trace "dropping fanout topic", topic
+  for topic in drops:
```
[the remark this answers is not in the document]
Comment (arnetheduck) [reply libp2p/protocols/pubsub/gossipsub/behavior.nim:470]: the old approach raised a `KeyError` - it's a downside of the `Table` API in general that there's no way to check that an item exists and if yes, use the value without the possibility of an exception

File: libp2p/protocols/pubsub/mcache.nim (hunk 1 above)
Comment (arnetheduck) [reply libp2p/protocols/pubsub/mcache.nim:30]: `mid in c.msgs` guarantees that the item is present, the try/except is just to silence the raises

**Comment by arnetheduck** [comment]
> Also, curious why are you using raiseAssert so aggressively, specially where it seems like a missing key is pretty innocuous?

`raiseAssert` in particular is `noreturn` meaning that if trees that end in it don't require a dummy expression to make the whole tree an expression - other than that, `Table` lacks a function that allows looking for and reading an element when it's present - `withValue` only works with mutable tables because it exposes a mutable item.

**Comment by arnetheduck** [comment]
> Nice cleanup

most importantly though, it fixes a significant scoring bug

**Pull request merged**